server/src/main/java/org/elasticsearch/cluster/coordination/StableMasterHealthIndicatorService.java

```diff
@@ -0,0 +1,433 @@
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.health.HealthIndicatorDetails;
+import org.elasticsearch.health.HealthIndicatorImpact;
+import org.elasticsearch.health.HealthIndicatorResult;
+import org.elasticsearch.health.HealthIndicatorService;
+import org.elasticsearch.health.HealthStatus;
+import org.elasticsearch.health.ImpactArea;
+import org.elasticsearch.health.SimpleHealthIndicatorDetails;
+import org.elasticsearch.health.UserAction;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.health.ServerHealthComponents.CLUSTER_COORDINATION;
+
+/**
+ * This indicator reports the health of master stability.
+ * If we have had a master within the last 30 seconds, and that master has not changed more than 3 times in the last 30 minutes, then
+ * this will report GREEN.
+ * If we have had a master within the last 30 seconds, but that master has changed more than 3 times in the last 30 minutes (and that is
+ * confirmed by checking with the last-known master), then this will report YELLOW.
+ * If we have not had a master within the last 30 seconds, then this will will report RED with one exception. That exception is when:
+ * (1) no node is elected master, (2) this node is not master eligible, (3) some node is master eligible, (4) we ask a master-eligible node
+ * to run this indicator, and (5) it comes back with a result that is not RED.
+ * Since this indicator needs to be able to run when there is no master at all, it does not depend on the dedicated health node (which
+ * requires the existence of a master).
+ */
+public class StableMasterHealthIndicatorService implements HealthIndicatorService, ClusterStateListener {
+
+    public static final String NAME = "stable_master";
+
+    private final ClusterService clusterService;
+    private final MasterHistoryService masterHistoryService;
+    /**
+     * This is the amount of time we use to make the initial decision -- have we seen a master node in the very recent past?
+     */
+    private final TimeValue veryRecentPast;
+    /**
+     * This is the number of times that it is OK for the master history to show a transition from a non-null master to a null master before
+     * it starts impacting the health status.
+     */
+    private final int acceptableNullTransitions;
+    /**
+     * This is the number of times that it is OK for the master history to show a transition one non-null master to a different non-null
+     * master before it starts impacting the health status.
+     */
+    private final int acceptableIdentityChanges;
```

Comment:
Same as before. Should we have this as a threshold so we don't have to +1 it everywhere?

Reply:
For this one we actually don't `+1` it anywhere, so it's probably good to stay as-is right?

Reply:
Ah, that's true. For consistency would it make sense to go with the threshold pattern and adapt the checks accordingly?
ie. have the condition 
```
if (masterChanges >= identityChangeThreshold) {
```

What do you think?

Reply:
I can do that.